### PR TITLE
docs(proposal): add EP-003 for VM CPU Power Model Server for Kepler

### PR DESCRIPTION
**Pull request by kaiyiliu-strala**
Introduces enhancement proposal for adding Machine Learning models to estimate kepler power metrics in a Virtual Machine environment when hardware power measurement interfaces like RAPL are not available.

### CODE REVIEW

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md [hunk 1]
```diff
@@ -0,0 +1,960 @@
...
+**Created**: 2025-08-25
+**Last Updated**: 2025-08-25
+
+## Summary
+
+This proposal outlines a comprehensive approach for training and deploying machine learning models to estimate CPU power consumption in virtual machine environments for Kepler energy monitoring. The system will train models using VM-accessible OS and memory counters as features, with baremetal Kepler QEMU process power measurements as targets, enabling accurate power estimation for package, core, DRAM, and uncore zones in virtualized environments where direct hardware power measurements are unavailable.
+
+## Problem Statement
+
+Virtual machines lack direct access to hardware power measurement interfaces (RAPL, IPMI, etc.) that are essential for energy monitoring in cloud and virtualized environments. Current Kepler deployments in VMs cannot provide accurate power consumption estimates because they cannot access the underlying hardware power consumption data. This creates a significant gap in energy monitoring capabilities for the growing virtualized infrastructure landscape.
```
Comment (vimalk78) [inline docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:14]: can we mention that PMU is usually disabled for VMs by providers, so VMs do not show any performance counter also.

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md [hunk 2]
```diff
@@ -0,0 +1,960 @@
...
+3. **Limited Power Visibility**: Cloud providers and VM operators lack visibility into their actual energy consumption
+4. **Inaccurate Estimations**: Existing VM power estimation methods are often generic and inaccurate for specific workloads
+5. **Deployment Complexity**: No standardized approach for VM-based energy modeling in Kepler ecosystem
+
+## Goals
+
+- **Primary Goal**: Develop zone-specific machine learning models (package, core, DRAM, uncore) for CPU power estimation in VMs
+- **Secondary Goal**: Create a production-ready deployment system for VM power models in Go environments
+- **Tertiary Goal**: Establish best practices for VM power modeling including CPU pinning and isolation requirements
+- **Performance Goal**: Achieve <10% mean absolute percentage error compared to baremetal measurements
```
Comment (vimalk78) [inline docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:29]: prefer RMS error over MAPE

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md [hunk 3]
```diff
@@ -0,0 +1,960 @@
...
+- Replacing baremetal RAPL measurements where available
+- Modeling power consumption for nested virtualization scenarios
+- Supporting real-time inference with <1ms latency requirements
+
+## Requirements
+
+### Functional Requirements
+
+- **FR1**: Train separate ML models for each power zone (package, core, DRAM, uncore)
+- **FR2**: Use only VM-accessible OS and memory counters as input features
```
Comment (vimalk78) [inline docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:45]: code seems to show it, but pls add a section showing list of features used for model training.

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md [hunk 4]
```diff
@@ -0,0 +1,960 @@
...
+kepler_vm_cpu_isolation_ratio{} 0.5
+kepler_vm_pinning_recommendation_score{} 0.75
+
+# Training and data quality metrics
+kepler_vm_training_data_points_total{zone="package"} 15420
+kepler_vm_model_accuracy_mae{zone="package"} 0.65
+kepler_vm_last_training_timestamp{zone="package"} 1692984532
+```
+
+## Implementation Plan
```
Comment (vimalk78) [inline docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:550]: typically enhancement proposals do not contain implementation plan.

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md
Comment (vimalk78) [inline docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:1]: please add some detail about

-  Model Architecture, mentioning model type (regression, neural network, xgboost etc.) with list of input features and feature engineering approach
- Training Data Requirements, mentioning training data source, if any preprocessing on data, training data size requirements
- how to detect/prevent model overfitting?
- hyperparameters


is there any dependence on number of vCPU? if not why, if yes what does this mean for model training/selection
consider a typical case of 128 CPU baremetal machine running multiple VMs, some with 4 vCPUs, some with 8 vCPUs

Comment (stephan-rayner) [reply docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:1]: > Model Architecture, mentioning model type (regression, neural network, xgboost etc.) with list of input features and feature engineering approach

I think this would be highly data dependent. Dimensionality is a concern of course, but also the linearity or non-linearity of the data is also import.
 
If the data is highly linear then a neural network or tree boosting approach like GBDT or GBM (xgboost wraps these) is over kill for our needs and some regression model would be effiencent. If the data contains minor non-linearities then a model NN or XGBoost may even still be over kill, because SVM with an RBF kernal could get you there and be more efficent (old school I know, but there is no school like the old school). If it is highly non-linear then the NN or XGBoost is totally sensible. Also if explainabilty is important then NNs would struggle there.

The other three all related to the first question and the data, so the short version is it depends and it may be too early to say with any confidence (especially if we don't have the EDA done)

**Comment by laurall974** [comment]
@vimalk78 @KaiyiLiu1234 @sunya-ch I think this PR can be closed.

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md (hunk 1 above)
Comment (bellalzohir) [inline docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md]: Could we consider an optional host-assisted telemetry mode for metrics that are invisible from inside the VM?

A collector running inside the guest cannot directly access hidden hardware metrics such as RAPL/package power, uncore activity, memory-controller behavior, CPU frequency states, or some PMU counters. However, the host or hypervisor could run a lightweight metric proxy associated with each VM and expose selected metrics to the guest through a controlled API, such as virtio/vsock, a guest-agent channel, a Prometheus endpoint, or a metadata-service-like interface.

The Kepler VM model could then support two modes:

1. guest-only mode, using only VM-visible features for maximum portability;
2. host-assisted mode, using additional proxy metrics for improved accuracy when the infrastructure allows host cooperation.

This may improve model accuracy while preserving compatibility with public-cloud cases where host-side telemetry is unavailable. The proposal should also discuss security, isolation, and multi-tenancy concerns, especially which host metrics are safe to expose and how they are mapped to a specific VM.

**Comment by sunya-ch** [comment]
Closed as https://github.com/sustainable-computing-io/kepler/pull/2464 merged.

@kaiyiliu-strala Please feel free to reopen if needed.

File: docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md
Comment (nikimanoledaki) [reply docs/developer/proposal/EP-003-VM-CPU-Power-Models-Training-Proposal.md:1]: referenced this in a separate issue for the model training itself: https://github.com/sustainable-computing-io/kepler/pull/2464
